src/rrd/mod.rs

```diff
@@ -0,0 +1,529 @@
+use anyhow::Result;
+use async_trait::async_trait;
+use core::time;
+use sqlx::{self, Sqlite};
+use std::{io::Write, path::Path};
+use thiserror::Error;
+
+#[derive(Debug, Error)]
+pub enum Error {
+    #[error("invalid window, can't be zero")]
+    InvalidWindow,
+
+    #[error("invalid retention, can't be zero or less than window")]
+    InvalidRetention,
+}
+
+/// Slot provides the functionality to set or overwrite the value of any metric
+/// at a specific timestamp.
+#[async_trait]
+pub trait Slot {
+    /// Counter sets (or overrides) the current stored value for this key,
+    /// with value
+    async fn counter(&mut self, key: &str, value: f64) -> Result<()>;
+    /// Key return the key of the slot which is the window timestamp
+    async fn key(&self) -> Result<i64>;
+}
+
+/// RRD is a round robin database of fixed size which is specified on creation
+/// RRD stores `counter` values. Which means values that can only go UP.
+/// then it's easy to compute the increase of this counter over a given window
+/// The database only keep history based on retention.
+#[async_trait]
+pub trait RRD<S, I, 'a>
+where
+    S: Slot,
+    I: Iterator,
+{
+    /// Slot returns the current window (slot) to store values.
+    async fn slot(&'a mut self) -> Result<S>;
+    /// Counters, return all stored counters since the given time (since) until now.
+    async fn counters(&self, since: std::time::SystemTime) -> Result<I>;
+    /// Last returns the last reported value for a metric given the metric
+    /// name
+    async fn last(&self, key: &str) -> Result<Option<f64>>;
+}
+
+/// SqliteRRD is the [`RRD`] implementation using Sqlite under the hood.
+pub struct SqliteRRD {
+    pool: sqlx::Pool<Sqlite>,
+    window: i64,
+    retention: i64,
+}
+
+/// SqliteSlot is the [`Slot`] implementation using Sqlite under the hood.
+pub struct SqliteSlot<'a> {
+    rrd: &'a mut SqliteRRD,
+    key: i64,
+}
+
+struct Counters {
+    index: usize,
+    inner: Vec<Counter>,
+}
+
+pub struct Counter {
+    metric: String,
+    value: f64,
+}
+
+impl Clone for Counter {
+    fn clone(&self) -> Self {
+        Counter {
+            metric: self.metric.clone(),
+            value: self.value,
+        }
+    }
+}
+
+impl Iterator for Counters {
+    type Item = Counter;
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.index == self.inner.len() {
+            return None;
+        }
+        let ret = self.inner[self.index].clone();
+        self.index += 1;
+        Some(ret)
+    }
+}
```

Comment:
There is no point of implementing an Iterator on top of a Vector since since Vector impelemnts IntoIterator. Iterators can be useful if the iteration actually drive the data retrieval from the data source.

When I said you can "even" return an iterator. I meant in sense that each call to "next()" on the iterator will actually advance the query cursor by "one" and get the data from the db. 

If you already retrieve everything in a vector already then what's the point of having an iterator. Just return the Vector. Or you return the vector as an iterator. 

for example like this
```rust
fn get() -> impl Iterator<Item = u64> {
    let data = vec![10, 20, 30];

    return data.into_iter();
}

fn main() {
    let x = get();
    for i in x {
        println!("{}", i);
    }
}
```
the get() function still return something that implements an Iterator. without actually implementing it. So your trait can still return an iterator while the implementation still loads everything in a Vector. Other implementation can implement a real iterator